docs/codeql/codeql-language-guides/abstract-syntax-tree-classes-for-working-with-ruby-programs.rst

```diff
@@ -0,0 +1,616 @@
+.. _abstract-syntax-tree-classes-for-working-with-ruby-programs:
+
+Abstract syntax tree classes for working with Ruby programs
+===========================================================
+
+CodeQL has a large selection of classes for representing the abstract syntax tree of Ruby programs.
+
+.. include:: ../reusables/abstract-syntax-tree.rst
+
+* An ``IDENTIFIER`` denotes an arbitrary identifier.
+* A ``CNAME`` denotes a class or module name.
+* An ``FNAME`` denotes a method name.
+* A ``TERM`` is either a semicolon or a newline used to terminate a statement.
+* Elements enclosed in ``« »`` are grouped and may be suffixed by ``?``, ``*``, or ``+`` to denote 0 or 1 occurrences, 0 or more occurrences, and 1 or more occurrences respectively.
+
+Statement classes
+~~~~~~~~~~~~~~~~~
+
+This table lists subclasses of Stmt_ representing Ruby statements.
+
++---------------------------------+--------------+----------------+---------+
+| Statement syntax                | CodeQL class | Superclasses   | Remarks |
++=================================+==============+================+=========+
+| ``alias`` FNAME FNAME           | AliasStmt_   | Stmt_          |         |
++---------------------------------+--------------+----------------+---------+
+| ``BEGIN {`` StmtSequence_ ``}`` | BeginBlock_  | StmtSequence_  |         |
++---------------------------------+--------------+----------------+---------+
+| ``begin`` StmtSequence_ ``end`` | BeginExpr_   | StmtSequence_  |         |
++---------------------------------+--------------+----------------+---------+
+| ``break`` «Expr_»?              | BreakStmt_   | ReturningStmt_ |         |
++---------------------------------+--------------+----------------+---------+
+| ``;``                           | EmptyStmt_   | Stmt_          |         |
++---------------------------------+--------------+----------------+---------+
+| ``END {`` StmtSequence_ ``}``   | EndBlock_    | StmtSequence_  |         |
++---------------------------------+--------------+----------------+---------+
+| ``next`` «Expr_»?               | NextStmt_    | ReturningStmt_ |         |
++---------------------------------+--------------+----------------+---------+
+| ``redo``                        | RedoStmt_    | Stmt_          |         |
++---------------------------------+--------------+----------------+---------+
+| ``retry``                       | RetryStmt_   | Stmt_          |         |
++---------------------------------+--------------+----------------+---------+
+| ``return`` «Expr_»?             | ReturnStmt_  | ReturningStmt_ |         |
++---------------------------------+--------------+----------------+---------+
+| ``undef`` «FNAME ``,``»+        | UndefStmt_   | Stmt_          |         |
++---------------------------------+--------------+----------------+---------+
+
+Calls
+~~~~~
+
```

Comment:
Should there be a brief introduction here?

Reply:
Yes, I'll add an intro here.